### ISSUE TITLE
Using llvmcall with VecElements

### ISSUE THREAD

**Issue by KristofferC** [eschnett/SIMD.jl#5]
Please tell me if you think this is the wrong place to discuss this.

I've been experimenting a little bit with using `llvmcall` on an immutable wrapper that wraps a tuple of `VecElements`.

The problem I have is that for `Vec`s of sizes that are not 2, 4, 8, LLVM gets the type info that this is a `[N x double]` instead of a `<N x double>` which causes the compilation to fail.
For sizes of 2, 4, 8 nice code is being generated and it even uses the AVX registers which is something that I hadn't got working without `llvmcall`

The code is here:

```jl
typealias LLVMDataTypeString ASCIIString
import Base.VecElement

d = Dict{DataType, LLVMDataTypeString}((
Float64 => "double",
Float32 => "float",
Int32 => "i32",
Int64 => "i64"
))

immutable Vec{N, T}
    data::NTuple{N, VecElement{T}}
end

@generated function Base.(:+){N, T}(x::Vec{N, T}, y::Vec{N, T})
    exp = """
    %3 = fadd <$(N) x $(d[T])> %0, %1
    ret <$(N) x $(d[T])> %3
    """
    return quote
            Vec(Base.llvmcall($exp,
               NTuple{N, VecElement{T}},
               Tuple{NTuple{N,VecElement{T}},NTuple{N,VecElement{T}}},
               x.data, y.data))
    end
end

F64 = VecElement(1.0)
F32 = VecElement(1.0f0)

# 128 bit xmm register used
@code_native Vec((F64, F64)) + Vec((F64, F64))

# 256 bit AVX register used
@code_native Vec((F64, F64, F64, F64)) + Vec((F64, F64, F64, F64))

# 2 256 bit AVX register vaddpd
@code_native Vec((F64, F64, F64, F64, F64, F64, F64, F64)) + 
             Vec((F64, F64, F64, F64, F64, F64, F64, F64))


# Erros because Vec((F64, F64, F64)) is interpreted as a [3 x double] instead
# of <3 x double>
@code_native Vec((F64, F64, F64)) + Vec((F64, F64, F64))
```

**Comment by eschnett** [comment]
How did you make this work??? All my attempts led to segfaults. I was waiting for @ArchRobison's fixes. Thanks for the example!

The restriction on the tuple size you mention is hard-coded into Julia. It's easy to remove there. Look for `julia_struct_to_llvm` in `cgutils.cpp`. There's an ominous undated warning about "LLVM assertion failures" that is quite possibly very outdated by now.

**Comment by KristofferC** [comment]
I actually do segfault when creating for example a `Vec((F64, F64, F64, F64))`  but presumably that can get fixed and the generated code will stay the same.

**Comment by eschnett** [comment]
I'm using
```
diff --git a/src/cgutils.cpp b/src/cgutils.cpp
index 45386e9..4338f71 100644
--- a/src/cgutils.cpp
+++ b/src/cgutils.cpp
@@ -1002,9 +1002,7 @@ static Type *julia_struct_to_llvm(jl_value_t *jt, bool *isboxed)
             }
             else {
                 if (isvector && lasttype != T_int1 && !type_is_ghost(lasttype)) {
-                    // TODO: currently we get LLVM assertion failures for other vector sizes
-                    bool validVectorSize = (ntypes == 2 || ntypes == 4 || ntypes == 8 || ntypes == 16);
-                    if (lasttype->isSingleValueType() && !lasttype->isVectorTy() && validVectorSize && is_vecelement_type(jl_svecref(jst->types,0)))
+                    if (lasttype->isSingleValueType() && !lasttype->isVectorTy() && is_vecelement_type(jl_svecref(jst->types,0)))
                         jst->struct_decl = VectorType::get(lasttype, ntypes);
                     else
                         jst->struct_decl = ArrayType::get(lasttype, ntypes);
```
This works fine, except for vector sizes that are a power of two minus one. Those result in a segfault, probably the same segfault as described above.

**Comment by KristofferC** [comment]
With "works fine" you mean that you can call `code_native` without getting segfaults. I still segfault when actually trying to create the `Vec`. Even as in my example above for a `<4 x double>` it segfaults.

I don't think the segfaults have only to do with the wrapper. Just generating a naked `(F64, F64, F64, F64)` segfaults for me.

**Comment by eschnett** [comment]
Yes, the segfault is caused by LLVM generating "aligned" memory load/store instructions. It should generate unaligned instructions instead, or Julia should ensure that the tuples actually are aligned.

I think you can avoid the segfaults by using length 1 and 2 vectors only, and avoiding length 4 vectors. Maybe length 4 vectors work if you stick with 32-bit datatypes.

**Comment by KristofferC** [comment]
> I think you can avoid the segfaults by using length 1 and 2 vectors only, 

That is pretty much exactly what I am doing here where I compute how many elements I should put in a `Vec`: https://github.com/KristofferC/SIMDVectors.jl/blob/5ca842326cb5755ec3ce09c619375eb54cef6c32/src/SIMDVector.jl#L35

Basically, as long as I put 128 bits in the `Vec` everything is fine. So yes, length 4 vectors works fine with 32 bit data types.

**Comment by KristofferC** [comment]
I tried the `vec_element` branch a bit and I get some spurious conversions when the result of a calculation is supposed to be stored back into the `Vec` type:

```
store <6 x double> %res.i, <6 x double>* %9, align 64
call void @julia_convert_31300([6 x double]* nonnull sret %4, %jl_value_t* inttoptr (i64 140130144794304 to %jl_value_t*), %Vec* nonnull %3) #0
```

It looks like julia is converting the vector to an array to store in the `Vec` object.

However, I did my own mini `Vec` implementation and then I don't get the conversion:

Here are two versions of some code: https://gist.github.com/KristofferC/6ba4bcb4951279aae2ea0e49e63b5d6b the upper one is with my mini `Vec` and the bottom one is using `SIMD.jl`. They do pretty much the same thing but only with `SIMD.jl` do I get the conversion code.

Any idea?
